google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1/Exceptions.java

```diff
@@ -0,0 +1,135 @@
+        @Nullable Throwable cause,
+        @Nullable String streamName,
+        ImmutableMap<String, GrpcStatusCode> errors) {
+      super(message, cause);
+      checkArgument(errors != null, "errors must be non null");
+      this.streamName = streamName;
+      this.errors = errors;
+    }
+
+    public ImmutableMap<String, GrpcStatusCode> getErrors() {
+      return errors;
+    }
+
+    public String getStreamName() {
+      return streamName;
+    }
+  }
+
+  /** Stream has already been finalized. */
+  public static final class StreamFinalizedException extends StorageException {
+    protected StreamFinalizedException(String name, String message, Throwable cause) {
+      super(message, cause, name, null);
+    }
+  }
+
+  /**
+   * There was a schema mismatch due to bigquery table with fewer fields than the input message.
+   * This can be resolved by updating the table's schema with the message schema.
+   */
+  public static final class SchemaMismatchedException extends StorageException {
+    protected SchemaMismatchedException(String name, String message, Throwable cause) {
+      super(message, cause, name, null);
+    }
+  }
+
+  private static StorageError toStorageError(com.google.rpc.Status rpcStatus) {
+    for (Any detail : rpcStatus.getDetailsList()) {
+      if (detail.is(StorageError.class)) {
+        try {
+          return detail.unpack(StorageError.class);
+        } catch (InvalidProtocolBufferException protoException) {
+          throw new IllegalStateException(protoException);
+        }
+      }
+    }
+    return null;
+  }
+
+  /**
+   * Converts a c.g.rpc.Status into a StorageException, if possible. Examines the embedded
+   * StorageError, and potentially returns a {@link StreamFinalizedException} or {@link
+   * SchemaMismatchedException} (both derive from StorageException). If there is no StorageError, or
+   * the StorageError is a different error it will return NULL.
+   */
+  @Nullable
+  public static StorageException toStorageException(
+      com.google.rpc.Status rpcStatus, Throwable exception) {
+    StorageError error = toStorageError(rpcStatus);
+    if (error == null) {
+      return null;
+    }
+    switch (error.getCode()) {
+      case STREAM_FINALIZED:
+        return new StreamFinalizedException(error.getEntity(), error.getErrorMessage(), exception);
+
+      case SCHEMA_MISMATCH_EXTRA_FIELDS:
+        return new SchemaMismatchedException(error.getEntity(), error.getErrorMessage(), exception);
+
+      default:
+        return null;
+    }
+  }
+
+  /**
+   * Converts a Throwable into a StorageException, if possible. Examines the embedded error message,
+   * and potentially returns a {@link StreamFinalizedException} or {@link SchemaMismatchedException}
+   * (both derive from StorageException). If there is no StorageError, or the StorageError is a
+   * different error it will return NULL.
+   */
+  @Nullable
+  public static StorageException toStorageException(Throwable exception) {
```

Comment:
Is there a way to convert the throwable back to a rpcStatus, I just realized that this is a string match.

Reply:
What type of Throwable do you have in mind? The problem is the Throwable only provides `getMesasge()` that contains information that specifically relates to `STREAM_FINALIZED` and `SCHEMA_MISMATCH_EXTRA_FIELDS` errors (unlike `AppendRowsResponse` which packs in error code and details). I don't think we can get away from the String match.

Reply:
Added a few TODO items to follow up when cl/408735437 is rolled out.